test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java

```diff
@@ -0,0 +1,142 @@
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.LockSupport;
+
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@State(Scope.Benchmark)
+@Fork(1)
+@Threads(1)
+@Warmup(iterations =  5, time = 5)
+@Measurement(iterations =  5, time = 5)
+public class UnparkBenchSleepersAfter {
+
+    /*
+        This micro creates thousands of sleeper threads after the threads doing the barrier await
+        to see if that has any effect on the barrier performance.
+     */
+
+    @Param({"4000"})
+    int idles;
+
+    @Param({"2"})
+    int workers;
+
+    CyclicBarrier barrier;
+
+    @Benchmark
+    public void barrier() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(workers);
+        for (int i = 0; i < workers; i++) {
+            exec.submit(() ->
+            {
+                try {
+                    barrier.await();
+                } catch (InterruptedException | BrokenBarrierException e) {
+                    barrier.reset();
+                } finally {
+                    latch.countDown();
+                }
+            });
+        }
+        latch.await();
+    }
+
+    IdleThread[] idle_threads;
+
+
+    ExecutorService exec;
+
+    @Setup
+    public void setup() throws InterruptedException {
+        barrier = new CyclicBarrier(workers);
+        exec = Executors.newFixedThreadPool(workers);
+        CountDownLatch latch = new CountDownLatch(workers);
+        for (int i = 0; i < workers; i++) { // warmup exec
+            exec.submit(() ->
+            {
+                try {
+                    Thread.sleep(0);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                } finally {
+                    latch.countDown();
+                }
+            });
+        }
+        latch.await();
+        idle_threads = new IdleThread[idles];
+        for(int i=0; i < idle_threads.length; i++) {
+            new Thread(idle_threads[i] = new IdleThread()).start();
+        }
+    }
+
+    @TearDown
+    public void tearDown() {
+        for(IdleThread it : idle_threads) {
+            it.stop();
+        }
+        exec.shutdown();
+    }
+
+    public static class IdleThread implements Runnable {
+        boolean done = false;
+        Thread my_thread;
+
+        @Override
+        public void run() {
+            my_thread = Thread.currentThread();
+            while (!done) {
```

Comment:
You might want to re-check IdleThread. From a quick look I would have expected "done" to be volatile. Also "my_thread" as it is set in the run with plain access.

Reply:
Yes `done` must be volatile. Surprised this even worked as expected.

Reply:
Same as we saw with `isAlive` before: `LockSupport.park` implicitly provides a compiler barrier. This `done` should be at least "opaque" to carry the same effect.

Reply:
Yes, `park()` and `unpark()` having release and acquire semantics is not documented [here](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/util/concurrent/locks/LockSupport.html#park()), and the documentation for `LockSupport` explicitly says "Reliable usage requires the use of volatile (or atomic) variables to control when to park or unpark. Orderings of calls to these methods are maintained with respect to volatile variable accesses, but not necessarily non-volatile variable accesses."

Reply:
Irrespective of park/unpark barriers, the JIT should have hoisted `done` and rewritten as:
```
if (Idone) {
  while(true) {
    ...
  }
}
```

Reply:
Well I guess the park/unpark barriers prevented that.